OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs

```diff
@@ -106,7 +106,7 @@ public EditorResourceLayer(Actor self, EditorResourceLayerInfo info)
 				kv => kv.Value.ResourceIndex,
 				kv => kv.Key);
 
-			Map.Resources.CellEntryChanged += UpdateCell;
+			((IMapResource)Map).Resources.CellEntryChanged += UpdateCell;
```

Comment:
A few lines above don't do `Map = self.World.Map;`. You have no idea whether that would give you a map implementation you can use. Check the type and only use if it implements `IMapResource`. Otherwise you are free to crash here so the modder knows not to use this trait with an incompatible map format.
Then remove all the casts like this one.
Repeat that for every affected trait in this commit.